pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -3882,18 +3885,139 @@ func TestChangefeedBareAvro(t *testing.T) {
 	cdcTest(t, testFn, feedTestForceSink("kafka"))
 }
 
+func toJSON(t *testing.T, x any) string {
+	t.Helper()
+	// Our json library formats differently than the stdlib, and json.MakeJSON()
+	// chokes on some inputs, so marshal it twice.
+	bs, err := gojson.Marshal(x)
+	require.NoError(t, err)
+	j, err := json.ParseJSON(string(bs))
+	require.NoError(t, err)
+	return j.String()
+}
+
 func TestChangefeedEnriched(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	key := map[string]any{"a": 0}
+	keySchema := map[string]any{
+		"name": "foo.key",
+		"fields": []map[string]any{
+			{"field": "a", "optional": false, "type": "int64"},
+		},
+		"optional": false,
+		"type":     "struct",
+	}
+
+	afterVal := map[string]any{"a": 0, "b": "dog"}
+	afterSchema := map[string]any{
+		"name":  "foo.after.value",
+		"field": "after",
+		"fields": []map[string]any{
+			{"field": "a", "optional": false, "type": "int64"},
+			{"field": "b", "optional": true, "type": "string"},
+		},
+		"optional": false,
+		"type":     "struct",
+	}
+	// TODO(#139662): add `before`
+
+	// Create an enriched source provider with no data. The contents of source
+	// will be tested in another test, we just want to make sure the structure &
+	// schema is right here.
+	esp := newEnrichedSourceProvider(changefeedbase.EncodingOptions{}, enrichedSourceData{})
+	source, err := esp.GetJSON(cdcevent.Row{})
+	require.NoError(t, err)
+
+	var sourceMap map[string]any
+	require.NoError(t, gojson.Unmarshal([]byte(source.String()), &sourceMap))
+
+	sourceSchema := esp.KafkaConnectJSONSchema().AsJSON()
+	var sourceSchemaMap map[string]any
+	require.NoError(t, gojson.Unmarshal([]byte(sourceSchema.String()), &sourceSchemaMap))
+	sourceSchemaMap["field"] = "source"
+
+	tsNsSchema := map[string]any{"field": "ts_ns", "optional": false, "type": "int64"}
+	opSchema := map[string]any{"field": "op", "optional": false, "type": "string"}
+
 	cases := []struct {
-		name               string
-		enrichedProperties []string
+		name                 string
+		enrichedProperties   []string
+		messageWithoutSource map[string]any
+		withSource           bool
+		expectedKey          map[string]any
 	}{
-		{name: "solo"},
-		{name: "with schema", enrichedProperties: []string{"schema"}},
-		{name: "with source", enrichedProperties: []string{"source"}},
-		{name: "with schema and source", enrichedProperties: []string{"schema", "source"}},
+		{
+			name: "with nothing",
+			messageWithoutSource: map[string]any{
+				// TODO: this is always the same now, so we can simplify these cases.
+				"after": afterVal,
+				"op":    "c",
```

Comment:
Wanted to check: `op` is always included if the user specifies `envelope=enriched`?